ironic/provider.go

```diff
@@ -90,8 +134,42 @@ func configureProvider(schema *schema.ResourceData) (interface{}, error) {
 		if err != nil {
 			return nil, fmt.Errorf("could not configure inspector endpoint: %s", err.Error())
 		}
-		clients.Inspector = inspector
+		clients.inspector = inspector
+	}
+
+	clients.timeout = schema.Get("timeout").(int)
+
+	return &clients, err
+}
+
+// Retries a gophercloud API until a timeout is reached.
+func waitForAPI(timeout int, client *gophercloud.ServiceClient) error {
+	var maxTries int
+	if timeout < 5 {
+		maxTries = 1
+	} else {
+		maxTries = timeout / 5
+	}
+	log.Printf("[DEBUG] Waiting for Ironic API to become available - max attempts is %d", maxTries)
+
+	httpClient := &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	for tries := 0; tries < maxTries; tries++ {
+		log.Printf("[DEBUG] Trying to connect to API, attempt %d of %d", tries, maxTries)
+		r, err := httpClient.Get(client.Endpoint)
+		if err == nil {
+			statusCode := r.StatusCode
+			r.Body.Close()
+			if statusCode == http.StatusOK {
+				log.Printf("[DEBUG] API successfully connected.")
+				return nil
+			}
+		}
+
+		time.Sleep(5 * time.Second)
```

Comment:
This function is not clever at all anymore, but we are still hitting a really weird problem on this sleep call.  If you call `time.Sleep` here, the Go scheduler never wakes up this goroutine again. It stays stuck forever.  The tests work fine, this only happens when running inside terraform in the installer.  There's a lot of goroutines (about 25) when this runs in the installer, but my box has GOMAXPROCS to 48, so there's not any resource starvation going on AFAICT.

I spent some time on the gopher slack and talked to some people who gave me suggestions about some debugging tools, to see what's going on (https://jvns.ca/blog/2017/09/24/profiling-go-with-pprof/).  I didn't see anything obvious other than seeing that the goroutine never wakes up again, it shows [sleep] next to it forever.

I'll keep on this next week, I'm on PTO tomorrow. I'm going to simplify the problem down to a bare bones terraform provider and see if any time.Sleep in terraform causes it to lock up, and then go from there.

The only workaround I can think of at the moment is a busy for loop instead of sleeping...but that's a terrible idea, TBH.

Reply:
It was not a concurrency issue.  time.Sleep was working, but after the first log gets printed, and the goroutine goes to sleep for n seconds, something is changing the log output to go to `ioutil.Discard`.  There's some suspicious looking code in Terraform, going to investigate more next week.